packages/js/email-editor/src/components/personalization-tags/rich-text-with-button.tsx

```diff
@@ -0,0 +1,138 @@
+import { BaseControl, Button } from '@wordpress/components';
+import { PersonalizationTagsModal } from './personalization-tags-modal';
+import { useCallback, useRef, useState } from '@wordpress/element';
+import {
+	createTextToHtmlMap,
+	getCursorPosition,
+	isMatchingComment,
+} from './rich-text-utils';
+import { useDispatch } from '@wordpress/data';
+import { useEntityProp } from '@wordpress/core-data';
+import { storeName } from '../../store';
+import { RichText } from '@wordpress/block-editor';
+import { __ } from '@wordpress/i18n';
+
+export function RichTextWithButton( {
+	label,
+	labelSuffix,
+	help,
+	placeholder,
+	attributeName,
+} ) {
+	const [ mailpoetEmailData ] = useEntityProp(
+		'postType',
+		'mailpoet_email',
+		'mailpoet_data'
+	);
+
+	const { updateEmailMailPoetProperty } = useDispatch( storeName );
+
+	const [ selectionRange, setSelectionRange ] = useState( null );
+	const [ isModalOpened, setIsModalOpened ] = useState( false );
+
+	const richTextRef = useRef( null );
+
+	const handleInsertPersonalizationTag = useCallback(
+		( value ) => {
+			// Retrieve the current value of the active RichText
+			const currentValue = mailpoetEmailData[ attributeName ] ?? '';
+
+			// Generate text-to-HTML mapping
+			const { mapping } = createTextToHtmlMap( currentValue );
+			// Ensure selection range is within bounds
+			const start = selectionRange?.start ?? currentValue.length;
+			const end = selectionRange?.end ?? currentValue.length;
+
+			// Default values for starting and ending indexes.
+			let htmlStart = start;
+			let htmlEnd = end;
+			// If indexes are not matching a comment, update them
+			if ( ! isMatchingComment( currentValue, start, end ) ) {
+				htmlStart = mapping[ start ] ?? currentValue.length;
+				htmlEnd = mapping[ end ] ?? currentValue.length;
+			}
+
+			// Insert the new tag
+			const updatedValue =
+				currentValue.slice( 0, htmlStart ) +
+				`<!--${ value }-->` +
+				currentValue.slice( htmlEnd );
+
+			// Update the corresponding property
+			updateEmailMailPoetProperty( attributeName, updatedValue );
+
+			setSelectionRange( null );
+		},
+		[
+			attributeName,
+			mailpoetEmailData,
+			selectionRange,
+			updateEmailMailPoetProperty,
+		]
```

Comment:
I see the useCallback needs quite a lot of deps. Looking at it I think we could make it more "cachable" if we passed some values as parameters. I think we could pass: the current value and selection range, and it would improve the caching.
So in onInsert, we would call

```js
handleInsertPersonalizationTag( value, mailpoetEmailData[ attributeName ] ?? '', selectionRange );
```

What do you think?